alerts/config.go

```diff
@@ -0,0 +1,129 @@
+package alerts
+
+// Data models for care team alerts.
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+// Config models a user's desired alerts.
+type Config struct {
+	Key string `json:"key" bson:"_id"`
+	// OwnerID links the user that owns and controls these alerts, i.e. the
+	// care team member.
+	OwnerID string `json:"ownerID"`
+	// InvitorID links the user whose data is shared, and will trigger alerts.
+	InvitorID string `json:"invitorID"`
+
+	UrgentLow       *WithThreshold `json:"urgentLow,omitempty"`
+	Low             *Deluxe        `json:"low,omitempty"`
+	High            *Deluxe        `json:"high,omitempty"`
+	NotLooping      *WithDelay     `json:"notLooping,omitempty"`
+	NoCommunication *WithDelay     `json:"noCommunication,omitempty"`
+}
+
+// Base describes the minimum specifics of a desired alert.
+type Base struct {
+	// Enabled controls whether notifications should be sent for this alert.
+	Enabled bool
+	// Repeat is measured in minutes.
+	Repeat DurationMinutes `json:"repeat"`
+}
+
+// DelayMixin adds a configurable delay.
+type DelayMixin struct {
+	// Delay is measured in minutes.
+	Delay DurationMinutes `json:"delay,omitempty"`
+}
+
+// ThresholdMixin adds a configurable threshold.
+type ThresholdMixin struct {
+	// Threshold is compared the current value to determine if an alert should
+	// be triggered.
+	Threshold `json:"threshold"`
+	// UnmarshalJSON prevents the use of the embedded Threshold.UnmarshalJSON.
+	UnmarshalJSON struct{}
+}
+
+// WithThreshold extends Base with ThresholdMixin.
+type WithThreshold struct {
+	Base
+	ThresholdMixin
+}
+
+// WithDelay extends Base with DelayMixin.
+type WithDelay struct {
+	Base
+	DelayMixin
+}
+
+// Deluxe extends Base with both DelayMixin and ThresholdMixin.
+type Deluxe struct {
+	Base
+	DelayMixin
+	ThresholdMixin
+}
+
+// DurationMinutes reads a JSON integer and converts it to a time.Duration.
+//
+// Values are specified in minutes.
+type DurationMinutes time.Duration
+
+func (m *DurationMinutes) UnmarshalJSON(b []byte) error {
+	if bytes.Equal(b, []byte("null")) || len(b) == 0 {
+		*m = DurationMinutes(0)
+		return nil
+	}
+	d, err := time.ParseDuration(string(b) + "m")
+	if err != nil {
+		return err
+	}
+	*m = DurationMinutes(d)
+	return nil
+}
+
+func (m DurationMinutes) Duration() time.Duration {
+	return time.Duration(m)
+}
+
+// ValueWithUnits binds a value to its units.
+//
+// Other types can extend it to parse and validate the Units.
+type ValueWithUnits struct {
+	Value float64 `json:"value"`
+	Units string  `json:"units"`
+}
+
+// Threshold is a value measured in either mg/dL or mmol/L.
+type Threshold ValueWithUnits
+
+// UnmarshalJSON adds validation of Units.
+func (t *Threshold) UnmarshalJSON(b []byte) error {
```

Comment:
A little odd to see validation logic as a custom unmarshaler. Doesn't this limit you to a single error per request? 

You may want to take a look at existing validation logic in the other platform services.

Reply:
> A little odd to see validation logic as a custom unmarshaler. 

Odd how? It makes sense to me. Part of unmarshaling is validation, both syntactically, and semantically.

> Doesn't this limit you to a single error per request?

No, though that is how I've coded it. I could still continue after receiving an error, grouping them up and returning a grouped error at the end. I find it's usually just simpler to return at the first error though. The code stays simpler that way.
 
> You may want to take a look at existing validation logic in the other platform services.

Will do. I looked around a bit before this, but clearly didn't look hard enough. Now that I know where some other validations are, I'll look at them more closely.

Reply:
> Odd how?

I don't think overloading marshaling/unmarshaling logic is a standard practice when working with web services, or at least I haven't seen this pattern in popular projects.

> Part of unmarshaling is validation, both syntactically, and semantically.

Marshaling deals with the way the data is formatted for transmission. I agree that there is validation step, e.g. when checking that a number fits in `int16`, but I think errors should only be raised when an incorrect data structure/primitive is used. Failing to unmarshal should mean that the data format is incorrect - e.g. when a client sends xml instead of json. 

Data validation comes after unmarshaling and ought to establish whether the received values lie within the semantic bounds we have defined.

> No, though that is how I've coded it. 

That's what I meant. Platform services usually return multiple errors - it's handy during development/debugging.